flink-runtime/src/main/java/org/apache/flink/runtime/memory/SharedResources.java

```diff
@@ -80,27 +81,37 @@
 		}
 	}
 
+	/**
+	 * Releases a lease (identified by the lease holder object) for the given type.
+	 * If no further leases exist, the resource is disposed.
+	 */
+	void release(String type, Object leaseHolder) throws Exception {
+		release(type, leaseHolder, (value) -> {});
+	}
+
 	/**
 	 * Releases a lease (identified by the lease holder object) for the given type.
 	 * If no further leases exist, the resource is disposed.
 	 *
-	 * @return True, if this was the last lease holder and the resource was disposed.
+	 * <p>This method takes an additional hook that is called when the resource is disposed.
 	 */
-	boolean release(String type, Object leaseHolder) throws Exception {
+	void release(String type, Object leaseHolder, Consumer<Long> releaser) throws Exception {
 		lock.lock();
 		try {
-			final LeasedResource resource = reservedResources.get(type);
+			final LeasedResource<?> resource = reservedResources.get(type);
 			if (resource == null) {
-				return false;
+				return;
 			}
 
 			if (resource.removeLeaseHolder(leaseHolder)) {
-				reservedResources.remove(type);
-				resource.dispose();
-				return true;
+				try {
+					reservedResources.remove(type);
+					resource.dispose();
+				}
+				finally {
+					releaser.accept(resource.size());
+				}
```

Comment:
I'm not sure whether calling the hook even when `resource.dispose` failed is a good idea. From the existing implementation of `SharedResources$LeasedResource#dispose`, it throws exception when `disposed` is `true` which means the resource is already disposed, following this logic the invocation of the hook may cause a memory "double release" ? Or if we assume `SharedResources$LeasedResource` will only be used with the lock (yes with current implementation), maybe we could remove the protection of multi-threading there as well as the exception declaration of the `dispose` method?

Reply:
I would like to keep the `finally` block, because of the following:
  - It would be possibly for the "release resource" call to fail
  - if we do not release the memory back, we must fail the process, because it leaved the TM in a "corrupt" state. It would not be possible to ever use that slot again.

Double releasing should not be possible, guarded by the code of `SharedResources`.

For clarity, I changed the `dispose()` method to not throw an exception on "double release".

Reply:
Would there be any chance that the resource is already disposed, but we call the hook twice? By "double releasing" I meant the accounting in `MemoryManager`.

Reply:
I guess my question is more related to the overall memory release logic than to this change.

What is the intended behaviour of Flink if `resource.dispose` fails? According to the contract `dispose` can throw an `Exception` and I assume it wouldn't be guaranteed that `resource` would have released all of its resources (native memory). 

If one follows the chain of close calls, then the release of the `SharedResources` could be triggered by `RocksDBKeyedStateBackend#dispose` where we swallow the exception via `IOUtils.closeQuietly`. Wouldn't this leave Flink in a problematic state where we have less memory than we think we have? I think it is dangerous to swallow the exceptions and it should be handled correctly. In the case of memory leakage I guess we should fail fatally.

Reply:
The double release of the "reservation" is actually guarded twice:
  - `if (resource.removeLeaseHolder(leaseHolder))`: Only the last release of a lease can trigger that.
  - The memory manager itself has this under an "owner", so if no memory is reserved under that owner, none can be released.